### ISSUE TITLE
rawAggregators as rawQuery?

### ISSUE THREAD

**Issue by bplsilva** [sksamuel/elastic4s#291]
Hello,
Is it somehow possible to enter a JSON with custom aggregators as rawAggregators, as it is possible to do for the body of the query?

Thanks.

**Comment by sksamuel** [comment]
Can you give me an example of what you mean as I'm not familar.

**Comment by bplsilva** [comment]
Oh, yes, of course. Sorry for not being very expressive before.

I actually want to run the following query with elastic4s and fetch the results:

```
{
    "query": {
        "filtered": {
            "filter": {
                "bool": {
                    "must": [
                        {
                            "term": {"categoryBidId": "47"}
                        },
                        {
                            "term": {"offerStatus": "Active"}
                        },
                        {
                            "term": {"algRef": "rank-label-based-nh"}
                        }
                    ]
                }
            }
        }
    },
    "aggs": {
        "by_sku_and_advertiser": {
            "terms": {
                "script": "doc['offerSku'].value + ':' + doc['advertiserId'].value",
                "order": {"by_labelsWeights>by_weightsSum" : "desc"}
            },
            "aggs" : {
                "by_labelsWeights" : {
                    "nested" : {
                        "path" : "labelsWeights"
                    },
                    "aggs" : {
                        "by_weightsSum" : {"sum": {"field" :"labelsWeights.weight"}}
                    }
                }
            }
        }
    }
}
```

It is straightforward to put the query part of the JSON above into a rawQuery and execute it as is.
It is however, as far as I know, impossible to enter the whole JSON, query + aggregators as a raw JSON into elastic4s and fetch the results.

I admit I have no clue on how to make this so far.
Is it possible? Is it something relatively easy to do?

Thanks!

**Comment by rakeshpant** [comment]
Were you able to achieve this? I also have a similar need. I want to use a script_file in range aggregation, but I couldn't find any option to do so.

**Comment by sksamuel** [comment]
I'm still a bit confused sorry - which bit do you want to use raw json for - the entire lot, or just the aggs / query bit ?

**Comment by rakeshpant** [comment]
I need to use raw json for aggs.

**Comment by sksamuel** [comment]
@rakeshpant  @bplsilva  You can do this now in master. Will be released as part of 1.5.12

Out of interest, why are you not using the aggregations DSL. Is it not powerful enough?

**Comment by bplsilva** [comment]
Hello @sksamuel,

Thank you very much for adding such a feature.
In my particular case, we already use the aggregations a lot in the project I work on, however there was a need from the people that generate and modify the queries to be able to define the aggregators dynamically, by simply modifying the json of the aggregations in the project.

I also have studied the aggregations DSL in depth after having asked my question, based on the samples in the project. We were not able to represent via DSL the aggregations of the example I have used in the first post, in the part of ordering stuff by a custom comparator:

```
"order": {"by_labelsWeights>by_weightsSum" : "desc"}
```
How would it be possible to do so using the aggregations DSL?

Anyway, thank you very much once again for this feature!

**Comment by ljmachado** [comment]
@sksamuel Is it possible to do something similar with the sort field? Like setting a rawSort JSON?

I could not figure it out by looking at how you did it with the aggregations.

**Comment by MLnick** [comment]
You can certainly order an aggregation by a custom comparator (at least in the sense I think you mean it) - here is an example from my code base that sets the order by a sub-aggregation using aggregation path:

```scala
val ordering = if (orderFrom) {
      Terms.Order.aggregation("denom>denom_count", reverseSortType)
    } else {
      Terms.Order.aggregation("numer>numer_count", reverseSortType)
    }
    val limit = if (num <= 1000) 1000 else num

    esClient.execute {
      search in indexName / docType query {
        filteredQuery.query(matchall).filter(termsFilterQ).filter(timeFilterQ)
      } searchType SearchType.Count aggs (
        agg terms aggName field key size limit order ordering aggs (
          agg filter "numer" filter termFilter(INTER_TYPE_FIELD, toType) aggs (agg count "numer_count" field TO_ID_FIELD),
          agg filter "denom" filter termFilter(INTER_TYPE_FIELD, fromType) aggs (agg count "denom_count" field TO_ID_FIELD)
          )
        )
    }
```

**Comment by Teudimundo** [comment]
We realized that the Aggregation accepting a raw jsonString has been removed when reorganizing the aggregations for some 5.x ES version. We would need to have it because we are injecting the aggregation part of a query that we are taking in input. Is there a new way to obtain the old behaviour?
Thanks.

**Comment by peterschrott** [comment]
@sksamuel, I know this is closed. 

How is it possible to pass the aggregation as a "raw json".

Thanks!

> @rakeshpant @bplsilva You can do this now in master. Will be released as part of 1.5.12
> 
> Out of interest, why are you not using the aggregations DSL. Is it not powerful enough?
